ddraw/IDirect3DDeviceX.cpp

```diff
@@ -338,15 +339,103 @@ HRESULT m_IDirect3DDeviceX::SetTransform(D3DTRANSFORMSTATETYPE dtstTransformStat
 			break;
 		}
 
-		HRESULT hr = (*d3d9Device)->SetTransform(dtstTransformStateType, lpD3DMatrix);
-
-		if (SUCCEEDED(hr))
+		if(Config.DdrawConvertHomogeneousW)
 		{
 #ifdef ENABLE_DEBUGOVERLAY
+			// Set the original matrix
 			DOverlay.SetTransform(dtstTransformStateType, lpD3DMatrix);
 #endif
+
+			if(dtstTransformStateType == D3DTS_VIEW)
+			{
+				D3DVIEWPORT9 Viewport9;
+				if(SUCCEEDED((*d3d9Device)->GetViewport(&Viewport9)))
+				{
+					const float width = (float)Viewport9.Width;
+					const float height = (float)Viewport9.Height;
+					const float ratio = width / height;
+
+					// Replace the matrix with one that handles D3DFVF_XYZRHW geometry
+					_D3DMATRIX view;
+					ZeroMemory(&view, sizeof(_D3DMATRIX));
+					view._11 = 2.0f / width;
+					view._22 = -2.0f / height;
+					view._33 = 1.0f;
+					view._41 = -1.0f;  // translate X
+					view._42 = 1.0f;   // translate Y
+					view._44 = 1.0f;
+
+					if(!Config.DdrawConvertHomogeneousToWorld)
+					{
+						// Override the original matrix
+						std::memcpy(lpD3DMatrix, &view, sizeof(_D3DMATRIX));
+					}
+					else
+					{
+						// Override the original matrix
+						std::memcpy(lpD3DMatrix, &view, sizeof(_D3DMATRIX));
```

Comment:
Just realized this basically ended up being the same code as above. So it can be moved out.

Reply:
Fixed this in the latest check in.